### PR TITLE
NEWSTACK-573: restore a single snapshot to multiple PVCs

### PR DESCRIPTION
**Pull request by david-enli**
**What this PR does / why we need it**:
this test specifically covers Apple's use case, create a single snapshot, and restore the snapshot to multiple volumes. The test will be disabled by default as it is time consuming, can be enabled with flag `PURE_FA_CLONE_MANY_TEST` set to true.

**Which issue(s) this PR fixes** (optional)
NEWSTACK-573

**Special notes for your reviewer**:

### CODE REVIEW

File: drivers/scheduler/k8s/k8s.go [hunk 1]
```diff
@@ -211,6 +211,7 @@ type K8s struct {
 	VaultToken              string
 	PureVolumes             bool
 	PureSANType             string
+	PureRunCloneMany        bool
```
Comment (dsupure) [inline drivers/scheduler/k8s/k8s.go:214]:
```suggestion
	RunCSISnapshotAndRestoreManyTest        bool
```

File: drivers/scheduler/k8s/k8s.go [hunk 2]
```diff
@@ -5231,6 +5233,59 @@ func (k *K8s) CSICloneTest(ctx *scheduler.Context, request scheduler.CSICloneReq
 	return nil
 }
 
+// CSISnapshotAndRestoreMany create CSI snapshot and restore to many PVCs, and we validate the PVCs are up and bound
+// (Testrail cases C58775, 58091)
+func (k *K8s) CSISnapshotAndRestoreMany(ctx *scheduler.Context, request scheduler.CSISnapshotRequest) error {
+	// This test will validate the content of the volume as opposed to just verify creation of volume.
+	if !k.PureRunCloneMany {
```
Comment (dsupure) [inline drivers/scheduler/k8s/k8s.go:5240]:
```suggestion
	if !k.RunCSISnapshotAndRestoreManyTest {
```

File: drivers/scheduler/k8s/k8s.go [hunk 3]
```diff
@@ -5231,6 +5233,59 @@ func (k *K8s) CSICloneTest(ctx *scheduler.Context, request scheduler.CSICloneReq
...
+
+	// creating the snapshot
+	volSnapshot, err := k.CreateCsiSnapshot(request.SnapName, pvcObj.Namespace, request.SnapshotclassName, pvcObj.Name)
+	if err != nil {
+		logrus.Errorf("Failed to create snapshot %s for volume %s", request.SnapName, pvcObj.Name)
+		return err
+	}
+
+	logrus.Infof("Successfully created snapshot: [%s] for pvc: %s", volSnapshot.Name, pvcObj.Name)
+	for i := 0; i < 500; i++ {
```
Comment (dsupure) [inline drivers/scheduler/k8s/k8s.go:5267]:
```suggestion
	for i := 0; i < numClones; i++ {
```

File: drivers/scheduler/k8s/k8s.go [hunk 4]
```diff
@@ -5231,6 +5233,59 @@ func (k *K8s) CSICloneTest(ctx *scheduler.Context, request scheduler.CSICloneReq
...
+			logrus.Errorf("Failed to build cloned PVC Spec: %s", err)
+			return err
+		}
+		_, err = k8sCore.CreatePersistentVolumeClaim(restoredPVCSpec)
+		if err != nil {
+			logrus.Errorf("Failed to restore PVC from snapshot %s: %s", volSnapshot.Name, err )
+			return err
+		}
+
+		if err = k.waitForPVCToBound(restoredPVCName, pvcObj.Namespace); err != nil {
```
Comment (dsupure) [inline drivers/scheduler/k8s/k8s.go:5280]: Let's avoid waiting here. Use the for loop to send requests parallel and have another for loop to check PVCs bound status later.

File: tests/common.go [hunk 5]
```diff
@@ -885,6 +898,41 @@ func ValidateCSIVolumeClone(ctx *scheduler.Context, errChan ...*chan error) {
 	})
 }
 
+// ValidatePureVolumeLargeNumOfClones is the ginkgo spec for restoring a snapshot to many volumes
+func ValidatePureVolumeLargeNumOfClones(ctx *scheduler.Context, errChan ...*chan error) {
+	context("For validation of an restoring large number from a snapshot", func() {
+		var err error
+		timestamp := strconv.Itoa(int(time.Now().Unix()))
+		snapShotClassName := PureSnapShotClass + "." + timestamp
+		if _, err := Inst().S.CreateCsiSanpshotClass(snapShotClassName, "Delete"); err != nil {
```
Comment (Pure-AdamuKaapan) [inline tests/common.go:907]: `CreateCsiSnapshotClass`

**Review by Pure-AdamuKaapan — COMMENTED** [review]
David pretty much hit all the requests I would have had too, so pretty much just +1 to everything he wrote

File: drivers/scheduler/k8s/k8s.go [hunk 6]
```diff
@@ -5231,6 +5239,64 @@ func (k *K8s) CSICloneTest(ctx *scheduler.Context, request scheduler.CSICloneReq
...
+		_, err = k8sCore.CreatePersistentVolumeClaim(restoredPVCSpec)
+		if err != nil {
+			logrus.Errorf("Failed to restore PVC from snapshot %s: %s", volSnapshot.Name, err )
+			return err
+		}
+
+	}
+	logrus.Info("Finished issueing PVC creation request, proceed to validate")
+
+	if err = k.waitForRestoredPVCsToBound(request.RestoredPVCName, pvcObj.Namespace); err != nil {
```
Comment (dsupure) [inline drivers/scheduler/k8s/k8s.go:5289]:
```suggestion
	if err = k.waitForRestoredPVCsToBound(request.RestoredPVCNamePrefix, pvcObj.Namespace); err != nil {
```

File: tests/common.go [hunk 7]
```diff
@@ -885,6 +898,41 @@ func ValidateCSIVolumeClone(ctx *scheduler.Context, errChan ...*chan error) {
 	})
 }
 
+// ValidatePureVolumeLargeNumOfClones is the ginkgo spec for restoring a snapshot to many volumes
+func ValidatePureVolumeLargeNumOfClones(ctx *scheduler.Context, errChan ...*chan error) {
+	context("For validation of an restoring large number from a snapshot", func() {
```
Comment (dsupure) [inline tests/common.go:903]:
```suggestion
	context("For validation of an restoring large number of volumes from a snapshot", func() {
```

**Review by dsupure — APPROVED** [review]
As long as you pass the FADA tests, please merge it! Good work.

**Pull request merged**